<div style="border:solid green 2px; padding: 20px">
    
<b>Олег, привет!</b> Мы рады тебя видеть на территории код-ревьюеров. Ты проделал большую работу над проектом, но давай познакомимся и сделаем его еще лучше! У нас тут своя атмосфера и несколько правил:


1. Меня зовут Александр Матвеевский. Я работаю код-ревьюером, моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных.
2. Общаемся на ты.
3. Если хочешь написать, спросить - не нужно стесняться. Только выбери свой цвет для комментария.  
4. Это учебный проект, тут можно не бояться сделать ошибку.  
5. У тебя неограниченное количество попыток для сдачи проекта.  
6. Let's Go!

---

Я буду красить комментарии цветом, пожалуйста, не удаляй их:

<div class="alert alert-block alert-danger">✍
    

__Комментарий от ревьюера №1__

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__


Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
</div>

---

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Такой комментарий  говорит о том, что было сделано что-то качественное и правильное =)
</div>
    
---
    
Предлагаю работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:   
    
<div class="alert alert-info"> <b>Комментарии студента:</b> Например, вот так.</div>
    
Всё это поможет выполнить повторную проверку твоего проекта оперативнее. Если будут какие-нибудь вопросы по моим комментариям, пиши, будем разбираться вместе :)    
    
---

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0  365286.622650         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              среднее             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        118514   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   
15   женат / замужем                 0      F   сотрудник     0        116820   
16   женат / замужем                 0      F   сотрудник     0        289202   
17  гражданский брак                 1      M   сотрудник     0         90410   
18    вдовец / вдова                 2      F   пенсионер     0         56823   
19         в разводе                 3      F   сотрудник     0        242831   

                                   purpose total_income_category  \
0                            покупка жилья                     B   
1                  приобретение автомобиля                     C   
2                            покупка жилья                     C   
3               дополнительное образование                     B   
4                          сыграть свадьбу                     C   
5                            покупка жилья                     B   
6                        операции с жильем                     B   
7                              образование                     C   
8                    на проведение свадьбы                     C   
9   

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

# IL - Installment Loan - кредит, который погашается в рассрочку.
# NPL - Non performing loans - просроченный кредит.
# sIL - доля кредитов для определённой категории относительно всех кредитов IL

# Создание сводной таблицы. столбцы:
# ||Кол-во детей||Выданные кредиты||Просроченные кредиты||Доля просроч.к выданным||
target = data.pivot_table(values='debt', index='children', aggfunc=['count','sum','mean'])\
    .droplevel(1, axis=1).reset_index()

# Переименование названий столбцов
target = target.rename(columns={'count': 'IL', 'sum': 'NPL', 'mean': 'NPL/IL'})

# Создание столбца 'sIL'
target['sIL'] = target['IL'] / sum(target['IL'])

# Сортировка по возрастанию категории
#target = target.sort_values(by='children', ascending=True)

# Форматирование для столбцов
target = target.style.format({'NPL/IL': '{:.2%}', 'sIL': '{:.1%}'})

display(target)


# ***Дополнительная проработка гипотезы. Суммирование категорий 2-5***

target2 = pd.DataFrame({"children": ["0", "1", "> 1"],
                        "IL"      : [0,     0,     0],
                        "NPL"     : [0,     0,     0],
                        "NPL/IL"  : [0.0, 0.0,   0.0],
                        "sIL"     : [0.0, 0.0,   0.0]})

target1 = data.pivot_table(values='debt', index='children', aggfunc=['count','sum','mean'])\
    .droplevel(1, axis=1)

for index, row in target1.iterrows():
    if index < 2:
        target2.loc[index, 'IL'] = row[0]
        target2.loc[index, 'NPL'] = row[1]
        target2.loc[index, 'NPL/IL'] = row[2]
    else:
        target2.loc[2, 'IL'] += row[0]
        target2.loc[2, 'NPL'] += row[1]

target2.loc[2, 'NPL/IL'] = target2.loc[2, 'NPL'] / target2.loc[2, 'IL']        

target2['IL'] = target2['IL'].astype('int')
target2['NPL'] = target2['NPL'].astype('int') 
target2['sIL'] = target2['IL'] / sum(target2['IL'])
target2 = target2.style.format({'NPL/IL': '{:.2%}', 'sIL': '{:.1%}'})    
 
display(target2)

Вывод: зависимость есть. Неполная.
Бездетные кредитуемые - лучшие клиенты банков! Остальные одинаково хуже.

Ниже статистически значимого уровня величины значений в категориях ['3','4','5'].

1) Если ['3','4','5'] отбросить, видно резкое увеличение доли "проблемных" кредитуемых просто при наличии детей. Увеличение есть и между категориями '1', '2' ребёнка, но незначительное - десятые доли процента.
2) Если суммировать категории ['3','4','5'] с категорией '2', разница между '1' и '> 1' ещё меньше - сотые доли процента.

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Здорово, что посчитал колличество заемщиков в каждой выборке, но тут стоит учесть неравнозначность выборок. 
    
---
    
Смотри  как эту таблицу можно вывести передав список агрегирующих функций:
    
```python
data.pivot_table(index = 'children', values = 'debt', aggfunc = ['count', 'sum', 'mean'])\
    .sort_values(by = ('mean', 'debt'), ascending = False)\
    .style.format({('mean', 'debt') : '{:.2%}'})
```
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
target = data.pivot_table(values='debt', index='family_status', aggfunc=['count','sum','mean'])\
    .droplevel(1, axis=1)
target = target.rename(columns={'count': 'IL', 'sum': 'NPL', 'mean': 'NPL/IL'})
target['sIL'] = target['IL'] / sum(target['IL'])
target = target.sort_values(by='NPL/IL', ascending=True)
target = target.reset_index()
target = target.style.format({'NPL/IL': '{:.2%}', 'sIL': '{:.1%}'})

display(target)

Вывод: зависимость есть. Семейные надёжнее "досемейных", а "послесемейные" надёжнее семейных.

Явно видна положительная "магическая" сила штампа в паспорте - сразу минус 1,7% (0,093-0,076) ненадёжных кредитуемых. Ещё лучше дела у вдовых и разведённых, но их гороздо меньше тех, кто ещё "в браке". А жаль.

Для банка очень желательно "неоштамповынных" сожительствующих клиентов организованно провести через ЗАГС.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

xDDDDDDD 
    
Верно. по расчетам можно сказать, что заемщики которые состоят или состояли в браке лучше оплачивают кредиты.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
target = data.pivot_table(values='debt', index='total_income_category', aggfunc=['count','sum','mean'])\
    .droplevel(1, axis=1)
target = target.rename(columns={'count': 'IL', 'sum': 'NPL', 'mean': 'NPL/IL'})
target['sIL'] = target['IL'] / sum(target['IL'])
target = target.sort_values(by='sIL', ascending=False)
target = target.reset_index()
target = target.style.format({'NPL/IL': '{:.2%}', 'sIL': '{:.1%}'})

display(target)

Вывод: зависимость есть. Прерывистая. Для двух основных категорий 'B' и 'C': выше доход - выше надёжность.

Исчезающе мало кредитов у "нищебродов" - "Е" и "богатеев" - "А". Наверное, первым мало дают, вторые мало берут.
Также ниже статистически значимого уровня количество кредитов у "преднищебродов" - "D".

Рассмотрим "В" и "С". Просрочка по кредитам чаще у "С" (доход 50 001-200 000), чем у "В" (доход 200 001-1 000 000), но "С" в три раза чаще берут кредиты.  

Пусть "богатеи" дадут деньги "нищебродам" и возьмут кредиты. Да побольше.

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Какую можно выявить зависимость, если сравнить две самые многочисленные категории?
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
target = data.pivot_table(values='debt', index='purpose_category', aggfunc=['count','sum','mean'])\
    .droplevel(1, axis=1)
target = target.rename(columns={'count': 'IL', 'sum': 'NPL', 'mean': 'NPL/IL'})
target['sIL'] = target['IL'] / sum(target['IL'])
target = target.sort_values(by='NPL/IL', ascending=True)
target = target.reset_index()
target = target.style.format({'NPL/IL': '{:.2%}', 'sIL': '{:.1%}'})

display(target)

Вывод: зависимость есть. Недвижимость в залоге - залог надёжности. Кредитуемого.

Надёжнее всего выдавать кредиты на недвижимость, чуть хуже на свадьбы и сильно менее надёжно на автомобиль и образование.

Для банка в качестве кредитуемого хуже потенциального автовладельца может быть только потенциальный автовладелец-студент-"платник".

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

xD в голос :D
    
Да, по расчетам можно сделать вывод, что самыми проблемными выглядят автокредиты и кредиты на образование.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Ответ: пропуски были в столбцах 'days_employed' и 'total_income'. В одинаковых строках, т.е. незаполнение данных в этих двух столбцах взаимосвязано.
Возможные причины (по снижению вероятности):
1) Не внесение данных ввиду их отсутствия, ошибки, намеренно и т.д.
2) Повреждение существующих данных.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Совершенно верно, человеческий фактор (кто-то что-то не ввел, умышленно или нет) также может стать причиной пропусков в данных.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Ответ: медианное значение максимально правдоподобно отражает выборку в целом.
Среднее арифметическое будет отражать ситуацию некорректно, в случае большого несимметричного выброса данных.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Верно, медиана менее чувствительна к выбросам в данных.
</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

По выдвинутым гипотезам зависимость в целом прослеживается. Не всегда полная, т.к. некоторые данные не имеют статистически значимую величину и не подходят для уверенного анализа.
Высоко влияние случайного распределения данных. Например, при проверке первой гипотезы: для категории "3-е детей" [330 27 0.082 1.5] достаточно на 2 просроченных кредита меньше для выхода на первое место; для "4-о детей" достаточно 1 кредита; для "5-о детей" 9 выданных и не просроченных кредитов опрокидывает всю статистику. Возможно имеет смысл группировка категорий ['2','3','4','5'] в одну - '> 1 ребёнка'.

Обобщая результаты проверок гипотез, можно сделать вывод о характерных чертах.

Люди, которые с наибольшей вероятностью допустят просрочку по кредиту:
1) Будут иметь 1-2 детей.
2) Возможно будут сожительствовать ("гражданский брак"), или не будут вовсе сожительствовать.
3) Их категория дохода - 'С' (доход 50 001-200 000).
4) Целью кредита будет получение образования или покупка автомобиля.

Комбинации из нескольких характерных черт могут указывать на возможные просрочки по кредиту.

Люди, которые с наибольшей вероятностью вернут кредит без просрочек:
1) Бездетные.
2) Состояли ранее, но более не состоит в браке.
3) Их категория дохода - 'В' (доход 200 001-1 000 000).
4) Целью кредита будет покупка недвижимости или проведение свадьбы.

Комбинации из нескольких характерных черт могут указывать на высокую надёжность кредитуемых.

    ***Рекомендованное дополнение***
Указанные выше выводы сделаны в результате следующих проделанных действий (кратко):
1) Импорт данных из 'https://code.s3.yandex.net/datasets/data.csv'.
2) Первичное изучение данных
    - количество столбцов*строк: 12*21525;
    - типы данных: float64(2), int64(5), object(5).
3) Предобработка данных: удаление пропусков - 4348, обработка аномальных значений - 2, 
4) Изменение типов данных - вещественный-->целочисленный.
5) Обработка дубликатов - 71.
6) Категоризация данных: для доходов и для целей кредита.
7) Проверка гипотез зависимости надёжности кредитуемых от:
    - количества детей на попечении - неполная (основные 3 из 5 категорий). Больше зависит от наличия детей;
    - семейного положения - полная (основные 5 из 5 категорий);
    - категории дохода - неполная (основные 2 из 5 категорий);
    - цели кредита - полная (4 из 4 категорий).
8) Ответы на тестовые вопросы - 2.
9) Общий вывод.

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Шикарный вывод и рекомендации для бизнеса. Кстати, хорошей практикой является написание в общем выводе всех твоих действий (кратко) по проекту. Ведь вывод раскрывает для заказчика всю твою проделанную работу и за что он платит деньги, даже если ему не интересен весь процесс исследования, то в отчёт он заглянет обязательно. Ты молодец!
</div>

<div style="border:solid blue 3px; padding: 20px">
<div class="alert alert-block alert-success">✔️
Комментарий от ревьюера №1 </b> 



У тебя получилась очень сильная и хорошая работа. Здорово, что не забываешь про комментарии, твой проект интересно проверять. 

---

Нужно поправить:

1) Ничего, отличная работа!)

----


Если у тебя будут какие-то вопросы по моим комментариям - обязательно пиши! Буду ждать работу на повторное ревью :)</div>